In [1]:
!pip install mo-gymnasium
!pip install git+https://github.com/LucasAlegre/morl-baselines.git

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 1.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 479.6/479.6 kB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 25.9 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
thinc 8.3.6 requires numpy<3.0.0,>=2.0.0, but you have numpy 1.26.4 which is incompatible.
  Cloning https://github.com/LucasAlegre/morl-baselines.git to /tmp/pip-req-build-ihzl5mbj
  Running command git clone --filter=blob:none --quiet https://github.com/LucasAlegre/morl-baselines.git /tmp/pip-req-build-ihzl5mbj
  Resolved https://github.com/LucasAlegre/morl-baselines.git to commit 92572aa5ee13e98b9e080f64df8e29602dba4f69
  Installin

In [1]:
import gymnasium as gym
import mo_gymnasium as mo_gym
import numpy as np
from MO_AWR import MO_AWR
#from google.colab import drive
#drive.mount('/content/drive')

In [2]:
env = mo_gym.make("deep-sea-treasure-v0")
agent = MO_AWR(env, log=False, td_lambda=.95, use_popf=False, max_buffer_size=1024, value_lr=1e-4, policy_lr=5e-4, batch_size=16, cd_threshold=0.2, beta=.5, l2_critic=0, use_critic_msbe=False, use_is_weighting=False)

agent.train(1000000, 512, num_value_steps=500, num_policy_steps=250, num_pf_points=20, num_expl_episodes=25, prio_sampling=False)

/usr/local/lib/python3.11/dist-packages/gymnasium/spaces/box.py:305: UserWarning: WARN: Box high's precision lowered by casting to float32, current high.dtype=float64
  gym.logger.warn(


Streaminguitvoer ingekort tot de laatste 5000 regels.
 [array([ 22.727, -21.206]) 20.9375]
 [array([ 22.35 , -18.723]) 18.3125]
 [array([ 4.44 , -2.972]) 2.875]
 [array([ 8.335, -3.014]) 3.0]
 [array([ 0.7, -1. ]) 1.0]
 [array([ 23.596, -27.591]) 29.9375]]
step 801002 	 return [ 20.356 -15.093], ([0.591 0.564]), new return [ 21.072 -14.627] 	 value loss 3.880E-02 	 policy loss 1.430E+00 	 popf loss 0.000E+00
Truncated eval episodes:  0
training value net
training policy
Pareto Front: [[array([ 20.779, -16.209]) 15.8125]
 [array([ 17.037, -11.026]) 10.8125]
 [array([ 20.292, -15.66 ]) 15.1875]
 [array([ 18.892, -13.134]) 12.9375]
 [array([ 18.965, -14.627]) 14.25]
 [array([ 17.701, -12.06 ]) 11.875]
 [array([ 20.912, -17.725]) 17.375]
 [array([ 9.787, -3.941]) 3.875]
 [array([ 23.465, -25.422]) 25.6875]
 [array([14.17 , -7.306]) 7.0625]
 [array([ 23.172, -23.129]) 22.6875]
 [array([ 2.411, -2.407]) 2.375]
 [array([11.006, -4.851]) 4.6875]
 [array([ 22.727, -21.206]) 20.9375]
 [array([ 2

In [3]:
!zip -r /content/results.zip /content/Results/

  adding: content/Results/ (stored 0%)
  adding: content/Results/training_plot_122.png (deflated 16%)
  adding: content/Results/training_plot_619.png (deflated 10%)
  adding: content/Results/training_plot_651.png (deflated 12%)
  adding: content/Results/training_plot_9.png (deflated 12%)
  adding: content/Results/training_plot_197.png (deflated 15%)
  adding: content/Results/training_plot_634.png (deflated 11%)
  adding: content/Results/training_plot_149.png (deflated 16%)
  adding: content/Results/training_plot_495.png (deflated 11%)
  adding: content/Results/training_plot_424.png (deflated 11%)
  adding: content/Results/training_plot_447.png (deflated 14%)
  adding: content/Results/training_plot_406.png (deflated 11%)
  adding: content/Results/training_plot_922.png (deflated 11%)
  adding: content/Results/training_plot_583.png (deflated 10%)
  adding: content/Results/training_plot_289.png (deflated 14%)
  adding: content/Results/training_plot_203.png (deflated 15%)
  adding: content/

In [12]:
!rm -rf /content/Results

In [11]:
import numpy as np
import torch as th
agent.experience_replay

[(-2e-05,
  204005,
  [Transition(observation=array([0, 0], dtype=int32), action=1, reward=array([ 0.7, -1. ], dtype=float32), return_=array([ 0.7, -1. ], dtype=float32), value_vec=array([ 18.15 , -19.275], dtype=float32), horizon=1, next_observation=array([1, 0], dtype=int32), action_prob=0.9217636, terminal=True)]),
 (-2e-05,
  204006,
  [Transition(observation=array([0, 0], dtype=int32), action=1, reward=array([ 0.7, -1. ], dtype=float32), return_=array([ 0.7, -1. ], dtype=float32), value_vec=array([ 18.15 , -19.275], dtype=float32), horizon=1, next_observation=array([1, 0], dtype=int32), action_prob=0.9217636, terminal=True)]),
 (-2e-05,
  204007,
  [Transition(observation=array([0, 0], dtype=int32), action=1, reward=array([ 0.7, -1. ], dtype=float32), return_=array([ 0.7, -1. ], dtype=float32), value_vec=array([ 18.15 , -19.275], dtype=float32), horizon=1, next_observation=array([1, 0], dtype=int32), action_prob=0.9217636, terminal=True)]),
 (-2e-05,
  204014,
  [Transition(observ

# Training Notes
- Most episodes in the ER only reach the first reward:
  - ER buffer is constructed such that new episodes are always added
  - This removes episodes by:
    1. Duplicates
    2. Distance metric (L2 + crowding)
    3. Oldest first
  - better returns are achieved as better PF returns are added to the PF

- buffer size, batch size and num expl episodes are important choices:
  - Num expl episodes needs to add enough episodes for them to be of significance during the next training iteration(s)
  - If num_exl is too large (and buffer size too small) then exploration episodes can overwrite episodes with good returns!
  - The smaller the batch size, the more these episodes have impact on training => faster convergence (depending on environment0
  - Larger buffer size means less chance of these exploration episodes to be used for training in the next timestep. A sufficient size is needed to ensure enough points are in te buffer for computing each point in the pf (at least batch_size*num_pf_points, but pref. more)
  - Preferably small batch size (16/32/64) and large buffer (1024/2048) depending on problem size
  - Depending on buffer size, pick sufficiently large number of exploration episodes to increase convergence speed

Idea:
- In order to improve the PF over time, average returns (from s0) must improve
- We need sufficient 'better' trajectories (at least batch_size) for this average to improve
- Take batches of (s,a) instead of only over state. This implies that taking similar actions in similar states leads to similar polices, which in term helps with convergence
- Using such batches could also solve loss/gradient issues!
- This complies with using this model for stochastic transitions, as we want averages over state action pairs
- This implies that we learn deterministic policies

Questions:
- Policy explorations improved drastically after using tanh inbetween hidden layers
- Value estimates are more accurate when using tanh
- Weighting losses using IS ratio should help because data is very off-policy compared to single-objective RL. => policy evaluation is still poor
  - High-valued returns are only achieved when randomness is used in the policy
- buffer returns get better, but eval returns are bad => all policies have high probability of instantly going down...

TODO:
- After policy evaluation it is clear that the agent often does not reach a terminal state => it learns to make cycles and uses these bootstrapped estimates to get its rewards...
  - weight policy loss so off-policy actions contribute less => less-likely to learn wrong actions by off policy learning... upon policy eval, the policies result in the first reward only
  - Experiment with Beta: slightly larger values result in better policy eval!

- Try to balance actor and critic updates:
  - In AWR, the actor is updated more than the critic. This is because the critic is only used to compute the advantage weights, and not for guiding the policy gradients themselves.
  - AWR remains stable without any target network for the critic for this same reason. In our case, value estimates are used for computing the TD(lambda) return (as in AWR). But this return is then used as conditioning for both the actor and critic.
- The POPF function behaves similar to the value function and is also trained using V: It might show similar overestimation issues!

- PF contains 'intermediate' returns: averages of different solutions which are non-dominated by themselves, but are not a distinct solution when evaluated.
  - This is problem specific...
  - Eg. the 3 returns next to each other are very close together (thus have a low crowding distance). Expected returns between 0.7 and 8.3 are in the PF, but not achievable (unless by stochastic policy). This is because crowding distance is higher than for the others.